core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4907,15 +4945,42 @@ public void rebuildDependencyGraph() {
      * @since 1.522
      */
     public Future<DependencyGraph> rebuildDependencyGraphAsync() {
-        dependencyGraphDirty.set(true);
+        synchronized (dependencyGraphLock) {
+            // Collects calls to this method to avoid unnecessary calculation of the dependency graph
+            if (scheduledFutureDependencyGraph != null) {
+                return scheduledFutureDependencyGraph;
+            }
+            // Schedule new calculation
+            return scheduledFutureDependencyGraph = scheduleCalculationOfFutureDependencyGraph(500, TimeUnit.MILLISECONDS);
+        }
+    }
+
+
+    private Future<DependencyGraph> scheduleCalculationOfFutureDependencyGraph(int delay, TimeUnit unit) {
         return Timer.get().schedule(() -> {
-            if (dependencyGraphDirty.get()) {
-                rebuildDependencyGraph();
+
+            //Wait for the currently running calculation to finish
+            while (calculatingFutureDependencyGraph != null && !calculatingFutureDependencyGraph.isDone()) {
+                Thread.sleep(100);
+            }
+
+            // Scheduled future becomes the currently calculating future
+            synchronized (dependencyGraphLock) {
+                calculatingFutureDependencyGraph = scheduledFutureDependencyGraph;
+                scheduledFutureDependencyGraph = null;
+            }
+
+            rebuildDependencyGraph();
+
+            // Mark that we finished calculating the dependency graph
+            synchronized (dependencyGraphLock) {
+                calculatingFutureDependencyGraph = null;
```

Comment:
We probably should not null this out here, we don't nullcheck this for any scheduling purpose.

Not nulling this will allow the above code to probably be simplified to
```
if (calculatingFutureDependencyGraph != null) {
    calculatingFutureDependencyGraph.get();
}
```

Reply:
I am now making sure that calculatingFutureDependencyGraph never becomes null. This made this null check obsolete and some other checks became obsolete as well and I deleted them. Thank you for your feedback.